sunbeam-python/sunbeam/plugins/observability/plugin.py

```diff
@@ -0,0 +1,455 @@
+OBSERVABILITY_DEPLOY_TIMEOUT = 1200  # 20 minutes
+CONTROLLER_MODEL = CONTROLLER_MODEL.split("/")[-1]
+
+
+class DeployObservabilityStackStep(BaseStep, JujuStepHelper):
+    """Deploy Observability Stack using Terraform"""
+
+    def __init__(
+        self,
+        plugin: "ObservabilityPlugin",
+        tfhelper: TerraformHelper,
+        jhelper: JujuHelper,
+    ):
+        super().__init__(
+            "Deploying Observability Stack", "Deploying Observability Stack"
+        )
+        self.tfhelper = tfhelper
+        self.jhelper = jhelper
+        self.model = OBSERVABILITY_MODEL
+        self.cloud = MICROK8S_CLOUD
+        self.read_config = lambda: plugin.get_plugin_info().get("config", {})
+        self.update_config = lambda c: plugin.update_plugin_info({"config": c})
+
+    def run(self, status: Optional[Status] = None) -> Result:
+        """Execute configuration using terraform."""
+
+        config = self.read_config()
+        tfvars = {
+            "model": self.model,
+            "cos-channel": "1.0/candidate",
+            "cloud": self.cloud,
+            "credential": f"{self.cloud}{CREDENTIAL_SUFFIX}",
+            "config": {"workload-storage": MICROK8S_DEFAULT_STORAGECLASS},
+        }
+        config.update(tfvars)
+        self.update_config(config)
+        self.tfhelper.write_tfvars(tfvars)
+
+        self.update_status(status, "deploying services")
+        try:
+            self.tfhelper.apply()
+        except TerraformException as e:
+            LOG.exception("Error deploying Observability Stack")
+            return Result(ResultType.FAILED, str(e))
+
+        apps = run_sync(self.jhelper.get_application_names(self.model))
+        LOG.debug(f"Application monitored for readiness: {apps}")
+        task = run_sync(update_status_background(self, apps, status))
+        try:
+            run_sync(
+                self.jhelper.wait_until_active(
+                    self.model,
+                    apps,
+                    timeout=OBSERVABILITY_DEPLOY_TIMEOUT,
+                )
+            )
+        except (JujuWaitException, TimeoutException) as e:
+            LOG.debug("Failed to deploy Observability Stack", exc_info=True)
+            return Result(ResultType.FAILED, str(e))
+        finally:
+            if not task.done():
+                task.cancel()
+
+        return Result(ResultType.COMPLETED)
+
+
+class DeployGrafanaAgentStep(BaseStep, JujuStepHelper):
+    """Deploy Grafana Agent using Terraform"""
+
+    def __init__(
+        self,
+        plugin: "ObservabilityPlugin",
+        tfhelper: TerraformHelper,
+        tfhelper_cos: TerraformHelper,
+        jhelper: JujuHelper,
+    ):
+        super().__init__("Deploying Grafana Agent", "Deploying Grafana Agent")
+        self.tfhelper = tfhelper
+        self.tfhelper_cos = tfhelper_cos
+        self.jhelper = jhelper
+        self.model = CONTROLLER_MODEL
+        self.read_config = lambda: plugin.get_plugin_info().get("config", {})
```

Comment:
This is overriding the TeraformVars for COS saved in cluster database. Use different keys for COS and grafana-agent terraform plans.